Defining Hyperdoctrine

GenericMonkey opened the pull request:
- Stole Eric's def'n of monotone functions and connected it to Cubical `Poset`. 
- Added notion of adjoint for monotone functions (same has HomBijection definition)
- Added `MonFunAdj` which is a monotone function with left and right adjoints
- Defined `POSET` and `POSETADJ`, categories with posets and monotone functions (w/ and w/o adjoints)
- Defined Hyperdoctrine from `Category` with `BinProds` to `POSETADJ` with Beck-Chevally Condition

maxsnew commented on Cubical/Categories/Instances/PosetCat/Base.agda:
```diff
@@ -0,0 +1,45 @@
+
+
+-- Category of Posets w/ Adjoints
+PosetCatAdjCycl : {ℓ ℓ' : Level} → Category
```
Probably worth defining a general subcategory construction that abstracts this at some point. Then it would be easier to define all 3 of monotone functions with left adj, right adj or both

maxsnew replied:
Universe levels should be explicit again

GenericMonkey replied:
Made the universe explicit, but think the subcategory construction is a more significant refactor that I might save for another PR. Looks like we'll need to rework it anyways with the new internal Heyting algebra stuff.

GenericMonkey replied:
Opened #32 to track this refactor.

maxsnew commented on Cubical/Categories/Constructions/Hyperdoctrine.agda:
```diff
@@ -0,0 +1,63 @@
+{-# OPTIONS --safe #-}
+
+module Cubical.Categories.Constructions.Hyperdoctrine where
```
I think this module shouldn't be `Hyperdoctrine` but instead something like `Hyperdoctrine.Unary` or `UnaryOverCartesian` because we will have other notions of hyperdoctrine as well.

In fact there are two axes of generalization: what structure we assume on the base category and what structure we assume about the fibers.

There's an even more basic version than the one defined here, which is that we could drop the requirement that the base category has products.

GenericMonkey replied:
Created a Hyperdoctrine directory and named this file Unary.agda.

maxsnew commented on Cubical/Categories/Constructions/Hyperdoctrine.agda:
```diff
@@ -0,0 +1,63 @@
+open import Cubical.Categories.Instances.PosetCat.Base
+open import Cubical.Categories.Instances.PosetCat.Monotone
+open import Cubical.Categories.Instances.PosetCat.MonotoneAdjoint
+open import UMP
```
I think we should maybe deprecate this module. I would instead use a category with chosen binary products

GenericMonkey replied:
Refactored to use the Limit.BinProduct with a field to assume the binproducts exist.

maxsnew commented on Cubical/Categories/Constructions/Hyperdoctrine.agda:
```diff
@@ -0,0 +1,63 @@
+    nat-left : ∀ {b c : cc .cat .ob}
+      (a : cc .cat .ob) (f : cc .cat [ b , c ]) →
+      ((MonComp
+        ((func ⟪ (prod₂-I cc a c {bin-prod-ob cc a b} (π₁ cc a b) ((π₂ cc a b) ⋆⟨ cc .cat ⟩ f)) ⟫) .morphism)
```
These definitions should look cleaner using `BinProduct.More.Notation`

GenericMonkey replied:
Finished the refactor with Limits.BinProduct

maxsnew commented on Cubical/Categories/Constructions/Hyperdoctrine.agda:
```diff
@@ -0,0 +1,63 @@
+                             ---------→
+                             left-adj
+
+    TODO: Can probably be more cleanly expressed as a bifunctor
```
I'll have to think about if there's a clean formulation of this property. I would guess the Kleisli category of the Env comonad is involved

maxsnew commented on Cubical/Categories/Instances/PosetCat/Monotone.agda:
```diff
@@ -0,0 +1,207 @@
+{-# OPTIONS --safe #-}
+
+-- TODO: This could be generalized to handle monotone functions on
+-- both preorders and posets.
```
Yeah I think my preference would be to define Preorder as the basic category and define Poset as a `Constructions.FullSubcategory`

GenericMonkey replied:
Opened #32 to track this refactor.

maxsnew commented on Cubical/Categories/Instances/PosetCat/Base.agda:
```diff
@@ -0,0 +1,45 @@
+open Category
+
+-- Category of Posets
+PosetCat : {ℓ ℓ' : Level} → Category
```
1. The universe levels should be explicit arguments
2. The naming convention in the library is to use ALLCAPS for these large categories so this should be `POSET ℓ ℓ'`. E.g. `SET` and `FUNCTOR`

GenericMonkey replied:
Fixed.

maxsnew commented on Cubical/Categories/Instances/PosetCat/Base.agda:
```diff
@@ -0,0 +1,45 @@
+
+-- Category of Posets w/ Adjoints
+PosetCatAdjCycl : {ℓ ℓ' : Level} → Category
+  (ℓ-max (ℓ-suc ℓ) (ℓ-suc ℓ')) (ℓ-max ℓ ℓ')
```
You can leave these in but in general Agda can infer these so you can just use underscores if you don't feel like figuring them out.

maxsnew commented on Cubical/Categories/Instances/PosetCat/Monotone.agda:
```diff
@@ -0,0 +1,207 @@
+
+
+-- If Y is a set, then functions into Y form a set
+isSetArrow : {X Y : Type} -> isSet Y -> isSet (X -> Y)
```
This is in the stdlib already, it's called `isSetΠ`

GenericMonkey replied:
removed.

maxsnew commented on Cubical/Categories/Instances/PosetCat/Monotone.agda:
```diff
@@ -0,0 +1,207 @@
+ -- Equivalence between MonFun' record and a sigma type   
+  unquoteDecl MonFun'IsoΣ = declareRecordIsoΣ MonFun'IsoΣ (quote (MonFun'))
+
+  Sigma : (X Y : Poset ℓ ℓ') -> Type (ℓ-max ℓ ℓ')
```
What is this for?

GenericMonkey replied:
I'm pretty sure that it's to define the isomorphism to a sigma type to get the `isSet` property nicely, but not really sure -- stole this file from the other repo.

maxsnew commented on Cubical/Categories/Instances/PosetCat/Monotone.agda:
```diff
@@ -0,0 +1,207 @@
+
+
+ -- Poset of monotone functions between two posets
+  IntHom : Poset ℓ ℓ' -> Poset ℓ ℓ' ->
```
Should refactor this in terms of `Exponential`

GenericMonkey replied:
Held off on this as it looks like we need to define the category of posets first before defining the exponential for it. Didn't want to create a circular dependency with the definition. We should probably move this to a different file to refactor

maxsnew commented on Cubical/Categories/Instances/PosetCat/MonotoneAdjoint.agda:
```diff
@@ -0,0 +1,105 @@
+
+module _ {ℓ ℓ' : Level} where
+
+  -- adjoints for monotone functions
```
Also called Galois connections. Should at least mention this in a comment for searchability

maxsnew commented on Cubical/Categories/Instances/PosetCat/MonotoneAdjoint.agda:
```diff
@@ -0,0 +1,105 @@
+    field
+      adjIff : ∀ {x y} → Iso
+        ((PosetStr._≤_ (Y .snd)) (MonFun.f L x) y)
+        ((PosetStr._≤_ (X .snd)) x (MonFun.f R y))
```
There's also a version of the universal element definition of adjoints for monotone functions and similar theorems to what yall did with representable profunctors. That stuff will probably be useful when we get to meets/joins/Heyting impl. We should think about if we can reuse the category stuff for that.

maxsnew reviewed (changes requested):
Looks good overall. Before merging the changes I'd like are the naming changes and avoid using UMP.agda. See the comments for details.

maxsnew commented on Cubical/Categories/Instances/PosetCat/Base.agda:
And I think the right name for this file would be `Instances.Posets.Base`

GenericMonkey replied:
Done

maxsnew commented:
Is this ready for review?

GenericMonkey commented:
Yes, I think so. Addressed the comments you had left about UMP and renaming, and opened an issue to track a future refactor of poset cats in terms of full subcategory and preorders.

GenericMonkey commented:
Will do a rebase once everything looks clean

maxsnew commented on Cubical/Categories/Constructions/Hyperdoctrine/Unary.agda:
```diff
@@ -41,23 +44,23 @@ record Hyperdoctrine ℓ ℓ' : Type (ℓ-suc (ℓ-max ℓ ℓ')) where
       ((MonComp
-        ((func ⟪ (prod₂-I cc a c {bin-prod-ob cc a b} (π₁ cc a b) ((π₂ cc a b) ⋆⟨ cc .cat ⟩ f)) ⟫) .morphism)
-        (fst ((func ⟪ π₂ cc a b ⟫) .left-adj)))
+        ((func ⟪ (cc .id) ×p f ⟫) .morphism)
```
much improved!

GenericMonkey commented:
Finished the rebase. I don't have permission to merge PRs though
